api/base/parsers.py

```diff
@@ -119,3 +119,35 @@ class JSONAPIParserForRegularJSON(JSONAPIParser):
     Allows same processing as JSONAPIParser to occur for requests with application/json media type.
     """
     media_type = 'application/json'
+
+class JSONAPIRelationshipParser(JSONParser):
+    """
+    Parses JSON-serialized data for relationship endpoints. Overrides media_type.
+    """
+    media_type = 'application/vnd.api+json'
+
+    def parse(self, stream, media_type=None, parser_context=None):
+        res = super(JSONAPIRelationshipParser, self).parse(stream, media_type, parser_context)
+
+        if not isinstance(res, dict):
+            raise ParseError('Request body must be dictionary')
+        data = res.get('data')
+
+        if data:
+            for value in data:
+
+                if value.get('id') is None:
+                    raise JSONAPIException(source={'pointer': '/data/id'}, detail=NO_ID_ERROR)
```

Comment:
The JSON pointers should contain the index of the invalid datum, e.g. `/data/3/id`).